rskj-core/src/main/java/co/rsk/peg/Federation.java

```diff
@@ -0,0 +1,152 @@
+    private List<BtcECKey> publicKeys;
+    private Instant creationTime;
+    private NetworkParameters btcParams;
+    private Script redeemScript;
+    private Script p2shScript;
+    private Address address;
+
+    public static Federation fromPrivateKeys(int numberOfSignaturesRequired, List<BtcECKey> privateKeys, Instant creationTime, NetworkParameters btcParams) {
+        List<BtcECKey> publicKeys = privateKeys.stream()
+                .map(key -> BtcECKey.fromPublicOnly(key.getPubKey()))
+                .collect(Collectors.toList());
+        return new Federation(numberOfSignaturesRequired, publicKeys, creationTime, btcParams);
+    }
+
+    public Federation(int numberOfSignaturesRequired, List<BtcECKey> publicKeys, Instant creationTime, NetworkParameters btcParams) {
+        this.numberOfSignaturesRequired = numberOfSignaturesRequired;
+        // Sorting public keys ensures same order of federators for same public keys
+        // Immutability provides protection unless unwanted modification, thus making the Federation instance
+        // effectively immutable
+        this.publicKeys = Collections.unmodifiableList(publicKeys.stream().sorted(BtcECKey.PUBKEY_COMPARATOR).collect(Collectors.toList()));
+        this.creationTime = creationTime;
+        this.btcParams = btcParams;
+        // Calculated once on-demand
+        this.redeemScript = null;
+        this.p2shScript = null;
+        this.address = null;
+    }
+
+    public List<BtcECKey> getPublicKeys() {
+        return publicKeys;
+    }
+
+    public int getNumberOfSignaturesRequired() {
+        return this.numberOfSignaturesRequired;
+    }
+
+    public Instant getCreationTime() {
+        return creationTime;
+    }
+
+    public NetworkParameters getBtcParams() {
+        return btcParams;
+    }
+
+    public Script getRedeemScript() {
+        if (redeemScript == null) {
+            redeemScript = ScriptBuilder.createRedeemScript(getNumberOfSignaturesRequired(), getPublicKeys());
+        }
+
+        return redeemScript;
+    }
+
+    public Script getP2SHScript() {
+        if (p2shScript == null) {
+            p2shScript = ScriptBuilder.createP2SHOutputScript(getNumberOfSignaturesRequired(), getPublicKeys());
+        }
+
+        return p2shScript;
+    }
+
+    public Address getAddress() {
+        if (address == null) {
+            address = Address.fromP2SHScript(btcParams, getP2SHScript());
+        }
+
+        return address;
+    }
+
+    @Override
+    public String toString() {
+        return String.format("%d of %d signatures federation", numberOfSignaturesRequired, publicKeys.size());
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        if (this == other) {
+            return true;
+        }
+
+        if (other instanceof Federation) {
+            return this.equalsFederation((Federation) other);
+        }
+
+        return false;
+    }
+
+    public boolean equalsFederation(Federation other) {
```

Comment:
Why do we need this two public methods instead of just the _equals_ from Object?